### PR TITLE
[UNF] Implement Comet, Stellar Pup

### PR DESCRIPTION
**Pull request by Susucre**
The Rules Text might be not quite right.

### CODE REVIEW

**Comment by Susucre** [comment]
[[Comet, Stellar Pup]]

File: Mage.Sets/src/mage/cards/c/CometStellarPup.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                return true;
+            }
+            player.choose(outcome, target, source, game);
+            Card card = game.getCard(target.getFirstTarget());
+            if (card != null) {
+                player.moveCards(card, Zone.HAND, source, game);
+            }
+        } else if (result <= 5) {
+            // Comet, Stellar Pup deals damage equal to the number of loyalty counters on him to a creature or player
+            TargetCreatureOrPlayer target = new TargetCreatureOrPlayer();
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/CometStellarPup.java:129]: .withNotTarget()

File: Mage.Sets/src/mage/cards/c/CometStellarPup.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            // [−2]
+            new RemoveCounterSourceEffect(CounterType.LOYALTY.createInstance(2))
+                    .apply(game, source);
+        } else {
+            //[+1]
+            new AddCountersSourceEffect(CounterType.LOYALTY.createInstance(1))
+                    .apply(game, source);
+
+            // You may activate Comet, Stellar Pup’s loyalty ability two more times this turn.
+            Permanent comet = game.getPermanent(source.getSourceId());
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/CometStellarPup.java:148]: source.getSourcePermanentIfItStillExists(game)

File: Mage.Sets/src/mage/cards/c/CometStellarPup.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                        new CometStellarPupContinousEffect(new MageObjectReference(comet, game)),
+                        source
+                );
+            }
+        }
+        return true;
+    }
+}
+
+class CometStellarPupContinousEffect extends ContinuousEffectImpl {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/CometStellarPup.java:160]: spelling: continuous

File: Mage.Sets/src/mage/cards/c/CometStellarPup.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+    private final MageObjectReference cometMOR;
+
+    public CometStellarPupContinousEffect(MageObjectReference cometMOR) {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/CometStellarPup.java:164]: package private

File: Mage.Sets/src/mage/cards/c/CometStellarPup.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        this.cometMOR = effect.cometMOR;
+    }
+
+    @Override
+    public CometStellarPupContinousEffect copy() {
+        return new CometStellarPupContinousEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/CometStellarPup.java:180]: does this comply with the ruling?

> In some unusual cases, Comet may gain loyalty abilities other than the one he normally has. In those cases, the effect of rolling a 6 applies only to his normal ability. (2022-10-07)

**Review by xenohedron — COMMENTED** [review]
looks fine, minor nits and a question

File: Mage.Sets/src/mage/cards/c/CometStellarPup.java (hunk 5 above)
Comment (Susucre) [reply Mage.Sets/src/mage/cards/c/CometStellarPup.java:180]: No, I did not see that ruling. Are [[Oath of Teferi]], [[The Chain Veil]] or [[Teferi, Master of Time]] not having that ruling due to being worded differently, and not being linked abilities?

It is more complex to make it so, involving a Watcher and permission specific to Comet's ability.

Comment (xenohedron) [reply Mage.Sets/src/mage/cards/c/CometStellarPup.java:180]: I don't mind leaving it as is

Comment (Susucre) [reply Mage.Sets/src/mage/cards/c/CometStellarPup.java:180]: Ok let me rebase and clean up.

File: Mage.Sets/src/mage/cards/c/CometStellarPup.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                    HasteAbility.getInstance(), Duration.EndOfTurn
+            ).setTargetPointer(new FixedTargets(token, game)), source);
+        } else if (result == 3) {
+            // [-1]
+            new RemoveCounterSourceEffect(CounterType.LOYALTY.createInstance(1))
+                    .apply(game, source);
+
+            // return a card with mana value 2 or less from your graveyard to your hand.
+            TargetCard target = new TargetCardInYourGraveyard(filterCard);
+            target.setNotTarget(true);
```
[the remark this answers is not in the document]
Comment (Susucre) [reply Mage.Sets/src/mage/cards/c/CometStellarPup.java:118]: notTarget is set to false if not using the 4 parameter constructor.
image

[the remark this answers is not in the document]
Comment (xenohedron) [reply Mage.Sets/src/mage/cards/c/CometStellarPup.java:118]: Uh sorry, got confused

[the remark this answers is not in the document]
Comment (Susucre) [reply Mage.Sets/src/mage/cards/c/CometStellarPup.java:118]: No worries, it is better to point out issues that are not real, than missing something important.
That's why I did double check too.

**Comment by Susucre** [comment]
Will add a few unit test.

File: Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java [hunk 7]
```diff
@@ -2196,11 +2196,15 @@ public boolean choose(Outcome outcome, Target target, Ability source, Game game,
             List<UUID> usedTargets = new ArrayList<>();
 
 
+            // TODO: Why TargetAnyTarget not handled there?
```
Comment (xenohedron) [inline Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java:2199]: The problem is that TargetCreatureOrPlayer doesn't extend TargetPermanentOrPlayer. TargetAnyTarget does.

`TargetCreatureOrPlayer` is used very rarely, just in Firesong and Sunspeaker and maybe some Star Wars cards, if I recall correctly.

I'd recommend refactoring it instead to inherit sensibly. That should maybe be a separate PR, and then rebase this one, unless you think it's clean and low risk change.

File: Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java [hunk 8]
```diff
@@ -2379,7 +2383,7 @@ public boolean choose(Outcome outcome, Target target, Ability source, Game game,
...
...````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````````
```
Comment (xenohedron) [inline Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java:2382]: that's a lot of backticks

File: Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java (hunk 7 above)
Comment (Susucre) [reply Mage.Tests/src/test/java/org/mage/test/player/TestPlayer.java:2199]: Oh that makes sense. I'll make a followup ticket for having `TargetCreatureOrPlayer` extend `TargetPermanentOrPlayer`.

The changes made here are quite safe, no other test was choosing a TargetCreatureOrPlayer, those tests would not have worked. Here I did the same as for TargetPermanentOrPlayer allowing to choose a permanent from the choice pool. (and player can not be choosen, which is also the same behavior)

**Review by xenohedron — APPROVED** [review]

**Pull request merged**